Mixed, saddle-point, div-grad problem using MFEM

nmnobre opened the pull request:
Refs #29030. Supersedes aurora-multiphysics/platypus#94.

nmnobre committed:
Add option to transpose mixed bilinear form kernels (#30637)

Co-authored-by: Bryce Shirley <bryce.shirley@stfc.ac.uk>

nmnobre committed:
Add divergence mbf kernel (#30637)

Co-authored-by: Bryce Shirley <bryce.shirley@stfc.ac.uk>

nmnobre committed:
Add flux integrated boundary condition (#30637)

Co-authored-by: Bryce Shirley <bryce.shirley@stfc.ac.uk>

nmnobre committed:
Add darcy example test (#30637)

Co-authored-by: Bryce Shirley <bryce.shirley@stfc.ac.uk>

nmnobre committed:
Add example index and descriptions (#29030)

Co-authored-by: Bryce Shirley <bryce.shirley@stfc.ac.uk>
Co-authored-by: Kingsley Collie <kingsley.collie@ukaea.uk>
Co-authored-by: Alexander Blair <alexander.blair@ukaea.uk>

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2912893/), step Docs: sync website on e44e4b0 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30637/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2915120/), step Generate coverage on 0577776 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>6c7356</th>
      <th colspan=4>#30637 057777</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.54%</td>
      <td>85.55%</td>
      <td>+0.01%</td>
      <td>90.32%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>113182</td>
      <td>113245</td>
      <td>+63</td>
      <td>56</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>19129</td>
      <td>19123</td>
      <td>-6</td>
      <td>6</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30637/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30637/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/30637/coverage/xfem)

</details>



This comment will be updated on new commits.

nmnobre committed:
Fix for correctly retrieving lfs for mixed problems (#30637)

nmnobre committed:
Fix destruction order for SuperLU solver and matrix (#30637)

Moose::MFEM::SuperLUSolver now both inherits and includes a member of type
mfem::SuperLUSolver. The former is so that MFEMSuperLU keeps the design of
all our other solvers that inherit from MFEMSolverBase, it wraps the
respective derived class of mfem::Solver. The latter is the solver we
actually use and is declared such that, on destruction of
Moose::MFEM::SuperLUSolver, the mfem::SuperLUSolver member is destroyed
before the mfem::SuperLURowLocMatrix member as assumed by mfem.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3018691/), step Clang format on f18957e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30637/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30637/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c6fc1730d3aee6171ba22f5cb89e182c5f1f564e`

nmnobre committed:
Fix PW material unit tests due to stricter attribute checks (#30808)

nmnobre committed:
Change getXXXCoefficient to take a parameter name (#30833)

The previous behavior, which used functions acting on proper names, is
still available under the more idiomatic getXXXCoefficientByName interface.

lindsayad commented:
> Refs #29030

An interesting reference 😄

nmnobre commented:
> > Refs #29030
> 
> An interesting reference 😄

It should've been https://github.com/idaholab/moose/discussions/29030#discussioncomment-13287679 to be more precise :)

lindsayad commented on framework/doc/content/syntax/MFEM/examples_index.md:
```diff
@@ -0,0 +1,41 @@
+# Examples that use MOOSE's MFEM capabilities
```
The "'" gets switched to " " on the a-z page. So does the "-" but I think "MFEM MOOSE" still looks fine compared to "MOOSE s MFEM" and it's consistent with other headings we use

nmnobre replied:
Changed to "MFEM" alone. The name "MFEM-MOOSE" hasn't grown on me yet, in a previous PR I've restricted it to that getting started doc. "MOOSE-MFEM" would perhaps sound a bit better, there's "MOOSE-NEML2" in a few places as well... Meh.

lindsayad commented on framework/include/mfem/kernels/MFEMVectorFEDivergenceKernel.h:
```diff
@@ -0,0 +1,32 @@
+#include "MFEMMixedBilinearFormKernel.h"
+
+/*
+(k div u, v)
```
```suggestion
 * (k div u, v)
```

nmnobre replied:
Done, and changed all other kernels at fault as well.

lindsayad commented on framework/src/mfem/userobjects/MFEMGeneralUserObject.C:
```diff
@@ -28,21 +28,39 @@ MFEMGeneralUserObject::MFEMGeneralUserObject(const InputParameters & parameters)
   return getMFEMProblem().getCoefficients().getMatrixCoefficient(name);
 }
 
+mfem::Coefficient &
+MFEMGeneralUserObject::getScalarCoefficient(const std::string & name)
+{
+  return getScalarCoefficientByName(getParam<MFEMScalarCoefficientName>(name));
+}
+
+mfem::VectorCoefficient &
+MFEMGeneralUserObject::getVectorCoefficient(const std::string & name)
+{
+  return getVectorCoefficientByName(getParam<MFEMVectorCoefficientName>(name));
+}
+
+mfem::MatrixCoefficient &
+MFEMGeneralUserObject::getMatrixCoefficient(const std::string & name)
+{
+  return getMatrixCoefficientByName(getParam<MFEMMatrixCoefficientName>(name));
+}
```
love it

lindsayad commented on test/tests/mfem/kernels/darcy.i:
```diff
@@ -0,0 +1,113 @@
+[Mesh]
+  type = MFEMMesh
+  file = ../mesh/star.mesh
+  uniform_refine = 4
```
how long does this test take?

lindsayad commented on test/tests/mfem/kernels/darcy.i:
```diff
@@ -0,0 +1,113 @@
+  [HDivFESpace]
+    type = MFEMVectorFESpace
+    fec_type = RT
+    fec_order = SECOND
+  []
+  [L2FESpace]
+    type = MFEMScalarFESpace
+    fec_type = L2
+    fec_order = SECOND
```
this is LBB stable with equal order?

nmnobre replied:
MFEM uses a different convention for the order of RT spaces: second, really means the space includes up to third order polynomials.

lindsayad commented on test/tests/mfem/kernels/darcy.i:
```diff
@@ -0,0 +1,113 @@
+    type = MFEMVectorL2Error
+    variable = velocity
+    function = exact_velocity
+    execution_order_group = 1
```
why is this needed?

nmnobre replied:
I thought it forced the postprocessors to execute after the solution had been obtained, so I left it.
But I've just removed those lines and the test still passed on my machine. Interestingly, I can't do the same for `irrotational.i` though...

lindsayad replied:
that's concerning. We should make an issue for this. This isn't necessary for the libMesh backend so we should try to setup things for the MFEM backend so it's not necessary there either

lindsayad commented on unit/src/MFEMMaterialTest.C:
why all these unit test changes?

nmnobre replied:
Because since #30808 our materials inherit from `MFEMBlockRestrictable` which, in its constructor, does some processing regarding the mesh attributes (i.e. blocks) the material applies to. We use `beam-tet.mesh` for these tests, which only has two blocks, "1" and "2", but we were setting material properties for a third block: `coef_params2.set<std::vector<SubdomainName>>("block") = {"3"};`.

nmnobre committed:
Fix doxygen /** detailed description blocks (#30637)

nmnobre commented on framework/src/mfem/bcs/MFEMVectorFEBoundaryFluxIntegratedBC.C:
```diff
@@ -0,0 +1,45 @@
+  InputParameters params = MFEMIntegratedBC::validParams();
+  params.addParam<MFEMScalarCoefficientName>(
+      "coefficient", "1.", "The coefficient which will be used in the integrated BC.");
+  return params;
```
Done. Well spotted, thank you!

nmnobre commented on framework/doc/content/syntax/MFEM/LinearElasticity.md:
```diff
@@ -0,0 +1,75 @@
+## Summary
+
+Solves a 3D linear elasticity problem for
+the deformation of a mult-material cantiliver beam. This example
```
Done. And "cantilever" is also mistyped :P

nmnobre commented on framework/include/mfem/bcs/MFEMVectorFEBoundaryFluxIntegratedBC.h:
```diff
@@ -0,0 +1,32 @@
+
+  MFEMVectorFEBoundaryFluxIntegratedBC(const InputParameters & parameters);
+
+  // Create MFEM integrator to apply to the RHS of the weak form. Ownership managed by the caller.
+  virtual mfem::LinearFormIntegrator * createLFIntegrator();
+
+  // Create MFEM integrator to apply to the LHS of the weak form. Ownership managed by the caller.
```
> Does this end up being owned by the MFEM lib?

Yes, once you add it to the respective (bi)linear form operator.

nmnobre commented on framework/include/mfem/bcs/MFEMVectorFEBoundaryFluxIntegratedBC.h:
```diff
@@ -0,0 +1,32 @@
+  MFEMVectorFEBoundaryFluxIntegratedBC(const InputParameters & parameters);
+
+  // Create MFEM integrator to apply to the RHS of the weak form. Ownership managed by the caller.
+  virtual mfem::LinearFormIntegrator * createLFIntegrator();
+
+  // Create MFEM integrator to apply to the LHS of the weak form. Ownership managed by the caller.
+  virtual mfem::BilinearFormIntegrator * createBFIntegrator();
```
For consistency will all the other classes, yes. Otherwise, probably not.

nmnobre commented on framework/include/mfem/bcs/MFEMVectorFEBoundaryFluxIntegratedBC.h:
```diff
@@ -0,0 +1,32 @@
+  // Create MFEM integrator to apply to the RHS of the weak form. Ownership managed by the caller.
+  virtual mfem::LinearFormIntegrator * createLFIntegrator();
+
+  // Create MFEM integrator to apply to the LHS of the weak form. Ownership managed by the caller.
```
Again, for consistency I don't really want to change this.
If I do, I'd want to change it throughout.

lindsayad replied:
I'd like to see it changed throughout then. It's a shame to miss out on these doc strings on the doxygen pages just because of one missing character

nmnobre commented on framework/doc/content/syntax/MFEM/Darcy.md:
```diff
@@ -0,0 +1,46 @@
+Solves a 2D mixed Darcy problem. This is a saddle point problem, discretized
+using Raviart-Thomas finite elements (velocity $\vec u$) and piecewise
+discontinuous polynomials (pressure $p$). This example demonstrates the use of
+transposition in the input file for mixed problems with different trial and
```
Help me here, what would you like to see?

nmnobre committed:
Fix doxygen /// brief/detailed descriptions (#30637)

nmnobre committed:
Fix documentation typos (#30637)

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

nmnobre committed:
Clarify the meaning of transposition (#30637)

nmnobre committed:
Prevent usage of inherited RecoverFEMSolution method (#30637)

moosebuild commented:
Job [Doxygen](https://civet.inl.gov/job/3056661/), step Framework doxygen on c3211f2 wanted to post the following:

View the framework doxygen [here](https://mooseframework.inl.gov/docs/PRs/30637/doxygen/moose/classes.html).

This comment will be updated on new commits.

moosebuild commented:
Job [Doxygen](https://civet.inl.gov/job/3056661/), step Modules doxygen on c3211f2 wanted to post the following:

View the modules doxygen [here](https://mooseframework.inl.gov/docs/PRs/30637/doxygen/modules/classes.html).

This comment will be updated on new commits.

nmnobre committed:
Fix out-of-bounds access in SymmetricRankTwoTensor::positiveProjectionEigenDecomposition (#30991)

nmnobre commented on framework/include/mfem/bcs/MFEMVectorFEBoundaryFluxIntegratedBC.h:
```diff
@@ -0,0 +1,32 @@
+
+  MFEMVectorFEBoundaryFluxIntegratedBC(const InputParameters & parameters);
+
+  // Create MFEM integrator to apply to the RHS of the weak form. Ownership managed by the caller.
```
Done throughout.

nmnobre committed:
Clarify usage of execution_order_group a bit further (#30637)

lindsayad commented on framework/doc/content/syntax/MFEM/examples_index.md:
```diff
@@ -38,4 +38,6 @@ that may serve as a useful starting point for users to adapt:
   problem, discretized using Raviart-Thomas finite elements (velocity $\vec u$) and
   piecewise discontinuous polynomials (pressure $p$). This example demonstrates the use of
   transposition in the input file for mixed problems with different trial and test variables.
+  Transposition flips the matrix associated a mixed bilinear form over its diagonal, swapping
+  the roles of the test and trial variables and avoiding the need to define a second kernel.
```
Oh. This should have been clear to me. I would've understand immediately if the word was "transpose"
